Fix cache functions inlining

carlocaione opened the pull request:
The current state of the cache framework doesn't allow for the full inlining of the cache functions. Try to fix this problem reworking the header files and move the ARM64 implementation to a fully inlined version by adding a new arch-specific cache-related header file. All the other architectures must do the same if they want to fully inline the cache operations.

carlocaione committed:
tests: cache: Add missing cases

Some test cases are missing. Add those back.

Signed-off-by: Carlo Caione <ccaione@baylibre.com>

carlocaione committed:
cache: Rework headers and force inlining

The current headers do not allow the full inlining of the cache
functions. Rework and cleanup the headers to allow for the full
inlining.

Signed-off-by: Carlo Caione <ccaione@baylibre.com>

carlocaione committed:
arm64: cache: Enable full inlining of the cache ops

Move all the cache functions to a standalone header to allow for the
inlining of these functions.

Signed-off-by: Carlo Caione <ccaione@baylibre.com>

stephanosio commented on include/zephyr/cache.h:
```diff
@@ -341,8 +257,10 @@ static inline int sys_cache_instr_flush_range(void *addr, size_t size)
  */
-__syscall int sys_cache_data_invd_range(void *addr, size_t size);
-static inline int z_impl_sys_cache_data_invd_range(void *addr, size_t size)
+__syscall int sys_cache_data_invd_range(void *addr, size_t size)
+	__attribute__((always_inline));
```
Should use `ALWAYS_INLINE` for portability.

carlocaione replied:
Yes, but that would confuse `scripts/build/parse_syscalls.py` that would be unable to parse anything (this is why the attribute is also put at the end). So, either we live with this or we have to modify the parsing script to account for `ALWAYS_INLINE`.

gmarull replied:
Sounds like a `parse_syscalls.py` issue that should be resolved.

carlocaione replied:
done, but be aware that my python-foo is not strong :D

stephanosio commented on include/zephyr/cache.h:
```diff
@@ -388,8 +306,10 @@ static inline int sys_cache_instr_invd_range(void *addr, size_t size)
  */
-__syscall int sys_cache_data_flush_and_invd_range(void *addr, size_t size);
-static inline int z_impl_sys_cache_data_flush_and_invd_range(void *addr, size_t size)
+__syscall int sys_cache_data_flush_and_invd_range(void *addr, size_t size)
+	__attribute__((always_inline));
```
same here

carlocaione replied:
ditto

stephanosio commented on include/zephyr/cache.h:
```diff
@@ -294,8 +208,10 @@ static inline int sys_cache_instr_flush_and_invd_all(void)
  */
-__syscall int sys_cache_data_flush_range(void *addr, size_t size);
-static inline int z_impl_sys_cache_data_flush_range(void *addr, size_t size)
+__syscall int sys_cache_data_flush_range(void *addr, size_t size)
+	__attribute__((always_inline));
```
ditto

carlocaione replied:
ditto

povergoing commented on include/zephyr/arch/cache.h:
```diff
@@ -0,0 +1,294 @@
+ * @{
+ */
+
+#if defined(CONFIG_ARM64)
```
Might there be too many `#if deifned()` for different `arch` in the future?

carlocaione replied:
well, it's a common pattern in Zephyr for arch headers: https://github.com/zephyrproject-rtos/zephyr/blob/c4565ad5b6527dd2474067230f98108518324ab4/include/zephyr/arch/cpu.h#L14-L36

povergoing replied:
Ah, right, forgot it's a header

carlocaione commented on scripts/build/parse_syscalls.py:
```diff
@@ -83,6 +83,15 @@ def analyze_headers(multiple_directories):
                         sys.stderr.write("Error decoding %s\n" % path)
                         raise
 
+                # Remove ALWAYS_INLINE to correctly parse fully-inlined
+                # syscalls defined as:
+                #
+                #      __syscall ALWAYS_INLINE sc(...)
+                #
+                # like it is done for some cache functions in
+                # include/zephyr/cache.h
+                contents = re.sub(r'ALWAYS_INLINE', '', contents)
```
yup, apparently also my regex-foo is not great :coffin:

still the CI is complaining about the `__syscall ALWAYS_INLINE` with `error: duplicate 'inline' declaration specifier `

carlocaione replied:
@stephanosio what about adding a new `__syscall_force_inline` instead?

stephanosio replied:
> @stephanosio what about adding a new `__syscall_force_inline` instead?

Sounds like a reasonable idea. Maybe `__syscall_always_inline` to align with the existing nomenclature.

carlocaione replied:
added

stephanosio commented on include/zephyr/toolchain/common.h:
```diff
@@ -143,8 +143,10 @@
  */
 #ifndef ZTEST_UNITTEST
 #define __syscall static inline
+#define __syscall_always_inline static inline __attribute__((always_inline))
 #else
 #define __syscall
+#define __syscall_always_inline
 #endif /* ZTEST_UNITTEST */
```
These changes (including to `parse_syscalls.py`) should go into a separate commit.

carlocaione commented on include/zephyr/toolchain/common.h:
```diff
@@ -143,8 +143,10 @@
  */
 #ifndef ZTEST_UNITTEST
 #define __syscall static inline
+#define __syscall_always_inline static inline __attribute__((always_inline))
 #else
 #define __syscall
+#define __syscall_always_inline
 #endif /* ZTEST_UNITTEST */
```
No, because I think this comment is also valid also for `__syscall_always_inline` https://github.com/zephyrproject-rtos/zephyr/blob/2236e31f6a49f9fb4f64cc1ebace010f1c0d86a9/include/zephyr/toolchain/common.h#L138-L142

carlocaione committed:
syscall: Introduce __syscall_always_inline

Sometimes we want to force the inlining of a __syscall. Introduce a new
__syscall_always_inline symbol to do that.

Signed-off-by: Carlo Caione <ccaione@baylibre.com>

carlocaione commented:
> whats up with docs failing? is this related to this pr?

no idea. I'm investigating.

carlocaione commented on include/zephyr/toolchain/common.h:
```diff
@@ -143,8 +143,10 @@
  */
 #ifndef ZTEST_UNITTEST
 #define __syscall static inline
+#define __syscall_always_inline static inline __attribute__((always_inline))
```
Nope, see https://github.com/zephyrproject-rtos/zephyr/pull/57021#discussion_r1172387821

ceolin replied:
That is not what I'm saying, I meant `#define __syscall_always_inline ALWAYS_INLINE`

ceolin reviewed (approved):
Looks much cleaner now :)
